core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java

```diff
@@ -30,6 +31,46 @@
     private final AtomicReference<V> resultRef = new AtomicReference<V>();
     /** Wait interval when the user calls get(). */
     private static final int GET_WAIT_INTERVAL_MILLIS = 1000;
+    /** Called when the future succeeds. */
+    private final Consumer<V> doneCallback;
+    /** Called when the future fails. */
+    private final Consumer<Exception> exceptionCallback;
+    /** Called when the future is canceled. */
+    private final Runnable canceledCallback;
+
+    /**
+     * Base constructor.
+     * @param doneCallback Called when the future succeeds.
+     * @param exceptionCallback Called when the future fails.
+     * @param canceledCallback Called when the future is cancelled.
+     */
+    public ImapFuture(final Consumer<V> doneCallback, final Consumer<Exception> exceptionCallback, final Runnable canceledCallback) {
```

Comment:
Please check out ImapFutureTest.java and add Unit Tests with use of your new ImapFuture(...) constructor.

You will need to test out the cases for done(), cancel() and those exception case

Reply:
any reason this need to be public? i would prefer this to be protected

Reply:
Constructor removed.